### PR TITLE
Unittest for Utilities and some other scenarios

### PR DESCRIPTION
**Pull request by ramigiki**

### CODE REVIEW

File: tests/apis/test_de_registration_reports_apis.py [hunk 1]
```diff
@@ -29,3 +29,35 @@
...
  POSSIBILITY OF SUCH DAMAGE.
 """
+
+
+import json
+import uuid
+import copy
+
+from tests._helpers import create_de_registration
+from tests.apis.test_de_registration_request_apis import REQUEST_DATA as DE_REG_REQ_DATA
```
[the remark this answers is not in the document]
Comment (a-wakeel) [reply tests/apis/test_de_registration_reports_apis.py]: Please check similar scenarios in your code

File: tests/apis/test_search_deregistraion_api.py [hunk 2]
```diff
@@ -66,6 +66,7 @@ def test_method_not_allowed(flask_app):
     data = json.loads(rv.data.decode('utf-8'))
     assert data.get('message') == 'method not allowed'
 
+
 def test_valid_search_specs(flask_app, db):
     seed_database(db)
     create_views(db)
```
Comment (a-wakeel) [inline tests/apis/test_search_deregistraion_api.py]: Can you please remove these as well? I think we missed it.
These methods should not be called in any test.

File: tests/db/test_registration_device.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+"""
+Status Model Unittests
```
Comment (a-wakeel) [inline tests/db/test_registration_device.py]: This is not Status Model right?

File: tests/db/test_registration_device.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from tests.apis.test_registration_request_apis import REQUEST_DATA as REG_REQ_DATA
+
+
+def test_sync_create(flask_app, app, db):  # pylint: disable=unused-argument
+    """ unittest for device creation async method"""
+
+    request = create_dummy_request(REG_REQ_DATA, 'Registration', status='Awaiting Documents')
+    response = Device.sync_bulk_create(request, 1, app)
+
+    # This method commits to db and returns nothing, need code change.
```
Comment (a-wakeel) [inline tests/db/test_registration_device.py]: So why are even we testing it now?

File: tests/misc/test_utilities.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
+"""
+DRS Package for DB unit tests
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: And these are not database unit tests right?

File: tests/misc/test_utilities.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+ OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+ TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import uuid
+import os
+from app.api.v1.helpers.utilities import Utilities
+from tests._helpers import create_registration
+from tests.apis.test_registration_request_apis import REQUEST_DATA as REG_REQ_DATA
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: ?

File: tests/misc/test_utilities.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    assert os.path.isdir(complete_path) is False
+
+
+def test_remove_file(app, session):  # pylint: disable=unused-argument
+    """Verify that the remove function works correctly."""
+    tracking_id = uuid.uuid4()
+    complete_path = os.path.join(app.config['DRS_UPLOADS'], '{0}'.format(tracking_id))
+    Utilities.create_directory(tracking_id)
+    assert os.path.isdir(complete_path)
+    Utilities.remove_file('file.txt', tracking_id)
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: How is it ensured that the file is removed? I mean this is just a call to the removal function and no assertion after?

Comment (a-wakeel) [reply tests/misc/test_utilities.py]: And which file is it removing?

File: tests/misc/test_utilities.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+def test_generate_summary(app, session, dirbs_dvs):  # pylint: disable=unused-argument
+    """Verify that the generate function works correctly."""
+
+    tracking_id = uuid.uuid4()
+    request = create_registration(REG_REQ_DATA, tracking_id)
+    Utilities.create_directory(tracking_id)
+    assert request.tracking_id == tracking_id
+    response = Utilities.generate_summary(IMEIS, tracking_id)
+    assert request.tracking_id == tracking_id
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: This is confusing, the assertion is only on the request.tracking_id (2 times), nothing on the response

File: tests/misc/test_utilities.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+def test_summary_status(app, session, dirbs_dvs):  # pylint: disable=unused-argument
+    """Verify that the generate function works correctly."""
+
+    tracking_id = uuid.uuid4()
+    request = create_registration(REG_REQ_DATA, tracking_id)
+    Utilities.create_directory(tracking_id)
+    assert request.tracking_id == tracking_id
+    # response = Utilities.check_summary_status('mock-task-id', request, app)
+    # assert request.tracking_id == tracking_id
+    # assert response is None
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: This is confusing as well, what is it simulating?

File: tests/misc/test_utilities.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+def test_pool_summary_request(app, session, dirbs_dvs):  # pylint: disable=unused-argument
+    """Verify that the generate function works correctly."""
+
+    tracking_id = uuid.uuid4()
+    request = create_registration(REG_REQ_DATA, tracking_id)
+    Utilities.create_directory(tracking_id)
+    assert request.tracking_id == tracking_id
+    response = Utilities.pool_summary_request('mock-task-id', request, app)
+    assert request.tracking_id == str(tracking_id)
+    assert response is None
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: This is not clear as well

File: tests/misc/test_utilities.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+
+def test_de_register_bulk_imeis(app, session):  # pylint: disable=unused-argument
+    """Verify that the de_register_imeis function works correctly."""
+
+    response = Utilities.de_register_imeis(IMEIS)
+    assert response is False
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: ??

File: tests/misc/test_utilities.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+
+def test_de_register_single_imeis(app, session):  # pylint: disable=unused-argument
+    """Verify that the de_register_imeis function works correctly."""
+
+    response = Utilities.de_register_imeis(IMEIS[0])
+    assert response is False
```
Comment (a-wakeel) [inline tests/misc/test_utilities.py]: ??

File: tests/_fixtures.py [hunk 13]
```diff
@@ -200,3 +202,56 @@ def dirbs_core(app):
...
     httpretty.reset()
+
+
+@pytest.yield_fixture(scope='session')
+def dirbs_dvs(app):
+    """
+    Mock server fixture to simulate DIRBS DVS API behaviours.
+    Monkey patch DIRBS-DVS calls made by DRS."""
+    httpretty.enable()
+    bulk_imei_response = {
```
[the remark this answers is not in the document]
Comment (ramigiki) [reply tests/_fixtures.py]: we only use bulk response in drs.

File: tests/misc/test_utilities.py (hunk 10 above)
Comment (ramigiki) [reply tests/misc/test_utilities.py]: This is async process so get no response in return.

File: tests/misc/test_utilities.py (hunk 11 above)
Comment (ramigiki) [reply tests/misc/test_utilities.py]: just checking an IMEI that is not inserted into database. It should return false.

File: tests/misc/test_utilities.py (hunk 12 above)
Comment (ramigiki) [reply tests/misc/test_utilities.py]: just checking an IMEI that is not inserted into database. It should return false.

File: tests/apis/test_de_registration_request_apis.py [hunk 14]
```diff
@@ -68,6 +68,24 @@ def test_request_get(flask_app, app, db, dirbs_core):  # pylint: disable=unused-
     assert data
 
 
+def test_request_get_single(flask_app, app, db, dirbs_core):  # pylint: disable=unused-argument
+    """ unittest for de-registration request"""
+    request = create_dummy_request(REQUEST_DATA, 'De-Registration', status='Awaiting Documents')
+
+    rv = flask_app.get("{0}/{1}".format(DEVICE_DE_REGISTRATION_REQ_API, request.id))
+    data = json.loads(rv.data.decode('utf-8'))
+    assert data
```
[the remark this answers is not in the document]
Comment (ramigiki) [reply tests/apis/test_de_registration_request_apis.py]: This has been incorporated to the code

File: tests/apis/test_de_registration_request_apis.py [hunk 15]
```diff
@@ -154,6 +172,48 @@ def test_request_update(flask_app, app, db, dirbs_core):  # pylint: disable=unus
...
+    headers = {'Content-Type': 'multipart/form-data'}
+    request_data = copy.deepcopy(REQUEST_DATA)
+    request_data['file'] = 'dummy-file.txt'
+    request = create_dummy_request(request_data, 'De-Registraiton', 'Pending Review')
+
+    request_data['device_count'] = 9
+    request_data['dereg_id'] = request.id
+
+    request_file = dict(request_data)
+    file_path = '{0}/{1}'.format('tests/unittest_data', 'de-registration_mock_file.txt')
```
[the remark this answers is not in the document]
Comment (ramigiki) [reply tests/apis/test_de_registration_request_apis.py]: This has been incorporated into the code.

File: tests/apis/test_search_deregistraion_api.py (hunk 2 above)
Comment (yasirz) [reply tests/apis/test_search_deregistraion_api.py]: > Can you please remove these as well? I think we missed it.
> These methods should not be called in any test.

cannot be, if i remove it, gives error message for foreign key violation.

File: tests/db/test_registration_device.py (hunk 4 above)
Comment (ramigiki) [reply tests/db/test_registration_device.py]: I have removed that code.

File: tests/apis/test_de_registration_reports_apis.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ramigiki) [reply tests/apis/test_de_registration_reports_apis.py]: This has been discussed with yasir. This is a todo. will be done in the next pull request.

**Pull request merged**